### PR TITLE
Basler acA1300

### PR DESCRIPTION
**Pull request by jangernert**
Still WIP

### CODE REVIEW

**Comment by f00f** [comment]
The camera series is actually named "Basler ace", so I would name the assembly, namespace, etc. accordingly.

File: BetaCameras/BaslerACA/BaslerACA.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+                {
+                    string msg = string.Format("BaslerACA: no device of type {0} found.", DeviceClass);
+                    log.Error(msg);
+                    throw new InvalidOperationException(msg);
+                }
+            }
+
+            _imageProvider.Open(dev.Index);
+
+            if (ActiveChannels.Count == 0)
```
Comment (f00f) [inline BetaCameras/BaslerACA/BaslerACA.cs]: Since this camera has only one channel I think it's safe to always activate it.
Otherwise, the correct pattern would be:
```
if (0 == ActiveChannels.Count)
{
    // activate default channels
    ActivateChannel(ChannelNames.Color);
}
else
{
    // run the activation code for channels the user has selected before Connect
    foreach (var channel in ActiveChannels)
    {
        ActivateChannelImpl(channel.Name);
    }
}
```

File: BetaCameras/BaslerACA/BaslerACA.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            }
+            else
+            {
+                dev = list.Where(d => d.Model.StartsWith(DeviceClass)).First();
+                if (null == dev)
+                {
+                    string msg = string.Format("BaslerACA: no device of type {0} found.", DeviceClass);
+                    log.Error(msg);
+                    throw new InvalidOperationException(msg);
+                }
```
Comment (f00f) [inline BetaCameras/BaslerACA/BaslerACA.cs]: Missing: read S/N

File: BetaCameras/BaslerACA/BaslerACA.cs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+            if (!String.IsNullOrEmpty(SerialNumber))
+            {
+                dev = list.Where(d => d.SerialNumber == SerialNumber).First();
+                if(null == dev)
+                {
+                    string msg = string.Format("BaslerACA: no device with s/n {0} found.", SerialNumber);
+                    log.Error(msg);
+                    throw new InvalidOperationException(msg);
+                }
```
Comment (f00f) [inline BetaCameras/BaslerACA/BaslerACA.cs]: Missing: read model name (for a generic implementation)

File: BetaCameras/BaslerACA/BaslerACA.cs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+namespace MetriCam2.Cameras
+{
+    public class BaslerACA : Camera
+    {
+        private const string DeviceClass = "acA1300";
+        private ImageProvider _imageProvider;
+        private ImageProvider.ImageReadyEventHandler _imageReady;
+        private Bitmap _bitmap;
+        private TaskCompletionSource<bool> _tsc = null;
```
[the remark this answers is not in the document]
Comment (f00f) [reply BetaCameras/BaslerACA/BaslerACA.cs]: Actually, I think an `AutoResetEvent` would be more suitable for the purpose

File: BetaCameras/BaslerACA/BaslerACA.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+            }
+            else
+            {
+                if (_bitmap != null)
+                {
+                    _bitmap.Dispose();
+                }
+
+                BitmapFactory.CreateBitmap(out _bitmap, image.Width, image.Height, image.Color);
+                BitmapFactory.UpdateBitmap(_bitmap, image.Buffer, image.Width, image.Height, image.Color);
```
Comment (f00f) [inline BetaCameras/BaslerACA/BaslerACA.cs]: This line is in both branches -> move it outside the conditional (then remove the if-branch)

File: BetaCameras/BaslerACA/BaslerACA.csproj [hunk 6]
```diff
@@ -0,0 +1,123 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="4.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="..\..\MetrilusReferencesVersions.props" />
+  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
+  <PropertyGroup>
+    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
```
Comment (f00f) [inline BetaCameras/BaslerACA/BaslerACA.csproj]: Default to x64 (there's no AnyCPU platform anyways)

Comment (f00f) [reply BetaCameras/BaslerACA/BaslerACA.csproj]: Or - if all references are AnyCPU - add the AnyCPU platform and remove all others.
(and uncheck "Prefer 32 bit")

File: BetaCameras/BaslerACA/Properties/AssemblyInfo.cs [hunk 7]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) Metrilus GmbH
+// MetriCam 2 is licensed under the MIT license. See License.txt for full license text.
+
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("MetriCam 2: CameraTemplate wrapper")]
```
Comment (f00f) [inline BetaCameras/BaslerACA/Properties/AssemblyInfo.cs]: Change

File: BetaCameras/BaslerACA/Properties/AssemblyInfo.cs [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+[assembly: AssemblyDescription("CHANGEME: MetriCam 2 wrapper for YOURCAMERAVENDOR cameras")]
```
Comment (f00f) [inline BetaCameras/BaslerACA/Properties/AssemblyInfo.cs]: Change

File: BetaCameras/BaslerACA/BaslerACA.cs [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+using System.Linq;
+using System.Drawing;
+using System.Threading.Tasks;
+using System.Collections.Generic;
+using PylonC.NETSupportLibrary;
+using PylonC.NET;
+
+namespace MetriCam2.Cameras
+{
+    public class BaslerACA : Camera
```
Comment (f00f) [inline BetaCameras/BaslerACA/BaslerACA.cs]: The Vendor property should be overwritten, the modelName set (in this case I'd set it to "Ace" in the c'tor, and then update it during Connect. Maybe also update it when DeviceClass (see comment below) is updated.

File: BetaCameras/BaslerACA/BaslerACA.cs [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+            if(_IsInitialized)
+            {
+                Pylon.Terminate();
+            }
+        }
+
+        public BaslerACA()
+        {
+            Init();
+            _imageReady = new ImageProvider.ImageReadyEventHandler(OnImageReadyEventCallback);
```
[the remark this answers is not in the document]
Comment (jangernert) [reply BetaCameras/BaslerACA/BaslerACA.cs]: I thought I would need to clean up the binding when the camera is disposed.

File: BetaCameras/BaslerACA/BaslerACA.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (jangernert) [reply BetaCameras/BaslerACA/BaslerACA.cs]: That's why this was marked as WIP. IDisposable was planed.

File: BetaCameras/BaslerACA/BaslerACA.cs [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+using System.Threading.Tasks;
+using System.Collections.Generic;
+using PylonC.NETSupportLibrary;
+using PylonC.NET;
+
+namespace MetriCam2.Cameras
+{
+    public class BaslerACA : Camera
+    {
+        private const string DeviceClass = "acA1300";
```
[the remark this answers is not in the document]
Comment (jangernert) [reply BetaCameras/BaslerACA/BaslerACA.cs]: DeviceClass is just a string used to filter the list of cameras that the Pylon library spits out. We would support all cameras when we don't use it at all. But since I'm unsure what the other cameras supported by Pylon are capable of, I check the camera name right now.

**Comment by f00f** [comment]
Ok, I've muted this PR. Notify me once it's review-ready.

**Comment by jangernert** [comment]
@f00f  ping!
was successfully tested in the Basler MetriX installation yesterday.
We would like to get this merged soon, so MetriX can reference MetriCam2 master builds again.

File: BetaCameras/BaslerACA/BaslerACA.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (f00f) [reply BetaCameras/BaslerACA/BaslerACA.cs]: Let me rephrase: "if it were configurable we would directly support all cameras compatible with acA1300"

[the remark this answers is not in the document]
Comment (jangernert) [reply BetaCameras/BaslerACA/BaslerACA.cs]: So make it a public property with the default value of "acA1300" so a developer can change it when he knows what he is doing, but GUI applications would still work only with tested cameras.
Probably makes sense to change this to a list of supported devices.
